### PR TITLE
ARROW-7754: [C++] Make Result<> faster

### PR DESCRIPTION
**Pull request by pitrou**
The main culprit for slowness was the underlying variant<> implementation.

The new implementation has comparable speed to returning Status + passing
a result-out pointer parameter.

### CODE REVIEW

**Comment by pitrou** [comment]
Benchmarks with gcc 7.4.
Before:
```
ErrorSchemeNoError                 6639 ns         6639 ns       317032 items_per_second=903.797M/s
ErrorSchemeBool                    6671 ns         6670 ns       315067 items_per_second=899.551M/s
ErrorSchemeStatus                  6691 ns         6690 ns       315033 items_per_second=896.894M/s
ErrorSchemeResult                 17563 ns        17561 ns       119493 items_per_second=341.658M/s
ErrorSchemeException               6661 ns         6660 ns       314578 items_per_second=900.89M/s
ErrorSchemeNoErrorNoInline         8055 ns         8054 ns       261182 items_per_second=745.009M/s
ErrorSchemeBoolNoInline            6966 ns         6965 ns       301415 items_per_second=861.476M/s
ErrorSchemeStatusNoInline         16561 ns        16558 ns       124946 items_per_second=362.354M/s
ErrorSchemeResultNoInline         26487 ns        26484 ns        79153 items_per_second=226.553M/s
ErrorSchemeExceptionNoInline       6713 ns         6713 ns       312970 items_per_second=893.829M/s
```
After:
```
ErrorSchemeNoError                 6625 ns         6624 ns       318577 items_per_second=905.814M/s
ErrorSchemeBool                    6627 ns         6626 ns       317117 items_per_second=905.542M/s
ErrorSchemeStatus                  6618 ns         6617 ns       317080 items_per_second=906.801M/s
ErrorSchemeResult                  6679 ns         6678 ns       315035 items_per_second=898.482M/s
ErrorSchemeException               6619 ns         6618 ns       317465 items_per_second=906.599M/s
ErrorSchemeNoErrorNoInline         7990 ns         7988 ns       261371 items_per_second=751.097M/s
ErrorSchemeBoolNoInline            6934 ns         6933 ns       303121 items_per_second=865.388M/s
ErrorSchemeStatusNoInline         16428 ns        16426 ns       127358 items_per_second=365.283M/s
ErrorSchemeResultNoInline         16525 ns        16522 ns       126809 items_per_second=363.152M/s
ErrorSchemeExceptionNoInline       7980 ns         7979 ns       263240 items_per_second=751.949M/s
```

**Comment by pitrou** [comment]
Benchmarks with clang 9.0.
Before:
```
ErrorSchemeNoError                 7201 ns         7200 ns       292201 items_per_second=833.287M/s
ErrorSchemeBool                    7270 ns         7269 ns       289349 items_per_second=825.452M/s
ErrorSchemeStatus                  6888 ns         6887 ns       303353 items_per_second=871.182M/s
ErrorSchemeResult                 48606 ns        48600 ns        43180 items_per_second=123.456M/s
ErrorSchemeException               7675 ns         7674 ns       273419 items_per_second=781.85M/s
ErrorSchemeNoErrorNoInline         8031 ns         8031 ns       261096 items_per_second=747.151M/s
ErrorSchemeBoolNoInline            7563 ns         7563 ns       278403 items_per_second=793.376M/s
ErrorSchemeStatusNoInline         16063 ns        16061 ns       130412 items_per_second=373.582M/s
ErrorSchemeResultNoInline         47883 ns        47878 ns        43879 items_per_second=125.319M/s
ErrorSchemeExceptionNoInline       7396 ns         7395 ns       283828 items_per_second=811.316M/s
```
After:
```
ErrorSchemeNoError                 7145 ns         7145 ns       294524 items_per_second=839.806M/s
ErrorSchemeBool                    7565 ns         7564 ns       277756 items_per_second=793.21M/s
ErrorSchemeStatus                  7094 ns         7093 ns       296323 items_per_second=845.898M/s
ErrorSchemeResult                 17428 ns        17424 ns       120180 items_per_second=344.344M/s
ErrorSchemeException               7606 ns         7605 ns       276110 items_per_second=788.972M/s
ErrorSchemeNoErrorNoInline         6631 ns         6630 ns       316267 items_per_second=904.932M/s
ErrorSchemeBoolNoInline            9286 ns         9285 ns       225941 items_per_second=646.179M/s
ErrorSchemeStatusNoInline         15925 ns        15923 ns       131750 items_per_second=376.802M/s
ErrorSchemeResultNoInline         17424 ns        17421 ns       120060 items_per_second=344.403M/s
ErrorSchemeExceptionNoInline       9358 ns         9357 ns       223584 items_per_second=641.213M/s
```

(for some reason, clang refuses to inline the `Result`-returning function...)

**Comment by pitrou** [comment]
@bkietz

File: cpp/src/arrow/result.h [hunk 1]
```diff
@@ -192,14 +207,27 @@ class Result : public util::EqualityComparable<Result<T>> {
   template <typename U, typename E = typename std::enable_if<
                             std::is_constructible<T, const U&>::value &&
                             std::is_convertible<U, T>::value>::type>
-  Result(const Result<U>& other) : variant_("unitialized") {
-    AssignVariant(other.variant_);
+  Result(const Result<U>& other) : status_(other.status_) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/result.h]: Does this cover the previous copy-constructor, e.g. if `U = T`?

File: cpp/src/arrow/result.h [hunk 2]
```diff
@@ -212,9 +240,15 @@ class Result : public util::EqualityComparable<Result<T>> {
   template <typename U,
             typename E = typename std::enable_if<std::is_constructible<T, U&&>::value &&
                                                  std::is_convertible<U, T>::value>::type>
-  Result(Result<U>&& other) : variant_("unitialized") {
-    AssignVariant(std::move(other.variant_));
-    other.variant_ = "Value was moved to another Result.";
+  Result(Result<U>&& other) noexcept {
+    if (ARROW_PREDICT_TRUE(other.status_.ok())) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/result.h]: Can you just:
```c++
std::swap(status_, other.status_);
std::swap(data_, other.data_);
```
That would remove the destructor crash issue.

File: cpp/src/arrow/result.h [hunk 3]
```diff
@@ -223,19 +257,30 @@ class Result : public util::EqualityComparable<Result<T>> {
...
-  Result& operator=(Result&& other) {
+  Result& operator=(Result&& other) noexcept {
     // Check for self-assignment.
     if (this == &other) {
       return *this;
     }
-    AssignVariant(std::move(other.variant_));
-    other.variant_ = "Value was moved to another Result.";
-
+    Destroy();
```
Comment (fsaintjacques) [inline cpp/src/arrow/result.h]: Ditto with `std::swap`, you wouldn't need any of the logic. Destructor of `other` would be properly cleanup upon going out of scope.

File: cpp/src/arrow/result_test.cc [hunk 4]
```diff
@@ -313,7 +313,7 @@ TYPED_TEST(ResultTest, MoveConstructorNonOkStatus) {
   Result<typename TypeParam::value_type> result2(std::move(result1));
 
   // Verify that the status of the donor object was updated.
-  EXPECT_FALSE(result1.ok());
+  //   EXPECT_FALSE(result1.ok());
```
Comment (fsaintjacques) [inline cpp/src/arrow/result_test.cc]: Testing leftover for this file?

File: cpp/src/arrow/result.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/result.h]: `std::swap(data_, other.data_);` will only work if `T` can be trivially moved using `memcpy`.

Comment (pitrou) [reply cpp/src/arrow/result.h]: And actually, it doesn't work at all, since `*this` isn't initialized at this point, so you're filling `other` with junk.

**Comment by pitrou** [comment]
I have tried using `__attribute__((always_inline))` and also passing `-mllvm -inline-threshold=10000`, and clang still doesn't seem to inline that `Result`-returning function. There is definitely something it chokes on.

It may actually be worthwhile trying to make a reproducer and opening an issue on clang?

File: cpp/src/arrow/result_test.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/arrow/result_test.cc]: Basically, the condition that was being tested is not true anymore. I'll remove the corresponding test lines.

**Comment by pitrou** [comment]
@bkietz Do you want to give this a review?

**Review by fsaintjacques — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @fsaintjacques  :-)

**Pull request closed without merging**